<a href="https://colab.research.google.com/github/ShirsaM/My-Google-Colab/blob/main/Data_Analysis_Importing_Datasets.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Data Acquisition
There are various formats for a dataset: .csv, .json, .xlsx etc. The dataset can be stored in different places, on your local machine or sometimes online.
In this section, you will learn how to load a dataset into our Jupyter Notebook.

In our case, the Automobile Dataset is an online source, and it is in a CSV (comma separated value) format. Let's use this dataset as an example to practice data reading.

In [1]:
# import pandas library
import pandas as pd
import numpy as np

## Read Data
We use pandas.read_csv() function to read the csv file. In the brackets, we put the file path along with a quotation mark so that pandas will read the file into a dataframe from that address. The file path can be either an URL or your local file address.
Because the data does not include headers, we can add an argument headers = None inside the read_csv() method so that pandas will not automatically set the first row as a header.

You can also assign the dataset to any variable you create.

In [2]:
# Import pandas library
import pandas as pd

# Read the online file by the URL provides above, and assign it to variable "df"
other_path = "https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DA0101EN-SkillsNetwork/labs/Data%20files/auto.csv"
df = pd.read_csv(other_path, header=None)

After reading the dataset, we can use the dataframe.head(n) method to check the top n rows of the dataframe, where n is an integer. Contrary to dataframe.head(n), dataframe.tail(n) will show you the bottom n rows of the dataframe.

In [3]:
# show the first 5 rows using dataframe.head() method
print("The first 5 rows of the dataframe") 
df.head(10)

The first 5 rows of the dataframe


0    1            2    3      4     5   ...    20   21    22  23  24     25
0   3    ?  alfa-romero  gas    std   two  ...   9.0  111  5000  21  27  13495
1   3    ?  alfa-romero  gas    std   two  ...   9.0  111  5000  21  27  16500
2   1    ?  alfa-romero  gas    std   two  ...   9.0  154  5000  19  26  16500
3   2  164         audi  gas    std  four  ...  10.0  102  5500  24  30  13950
4   2  164         audi  gas    std  four  ...   8.0  115  5500  18  22  17450
5   2    ?         audi  gas    std   two  ...   8.5  110  5500  19  25  15250
6   1  158         audi  gas    std  four  ...   8.5  110  5500  19  25  17710
7   1    ?         audi  gas    std  four  ...   8.5  110  5500  19  25  18920
8   1  158         audi  gas  turbo  four  ...   8.3  140  5500  17  20  23875
9   0    ?         audi  gas  turbo   two  ...   7.0  160  5500  16  22      ?

[10 rows x 26 columns]

In [4]:
print("The last 10 rows are: ")
df.tail(5)

The last 10 rows are: 


0   1      2       3      4     5   ...    20   21    22  23  24     25
200  -1  95  volvo     gas    std  four  ...   9.5  114  5400  23  28  16845
201  -1  95  volvo     gas  turbo  four  ...   8.7  160  5300  19  25  19045
202  -1  95  volvo     gas    std  four  ...   8.8  134  5500  18  23  21485
203  -1  95  volvo  diesel  turbo  four  ...  23.0  106  4800  26  27  22470
204  -1  95  volvo     gas  turbo  four  ...   9.5  114  5400  19  25  22625

[5 rows x 26 columns]

## Add Headers
Take a look at our dataset. Pandas automatically set the header with an integer starting from 0.

To better describe our data, we can introduce a header. This information is available at: https://archive.ics.uci.edu/ml/datasets/Automobile.

Thus, we have to add headers manually.

First, we create a list "headers" that include all column names in order. Then, we use dataframe.columns = headers to replace the headers with the list we created.

In [6]:
# create headers list
headers = ["symboling","normalized-losses","make","fuel-type","aspiration", "num-of-doors","body-style",
         "drive-wheels","engine-location","wheel-base", "length","width","height","curb-weight","engine-type",
         "num-of-cylinders", "engine-size","fuel-system","bore","stroke","compression-ratio","horsepower",
         "peak-rpm","city-mpg","highway-mpg","price"]

df.columns = headers
df.head(5)

symboling normalized-losses         make  ... city-mpg highway-mpg  price
0          3                 ?  alfa-romero  ...       21          27  13495
1          3                 ?  alfa-romero  ...       21          27  16500
2          1                 ?  alfa-romero  ...       19          26  16500
3          2               164         audi  ...       24          30  13950
4          2               164         audi  ...       18          22  17450

[5 rows x 26 columns]

**We need to replace the "?" symbol with NaN so the dropna() can remove the missing values:**

In [7]:
df1=df.replace('?',np.NaN)


**We can drop missing values along the column "price" as follows:**

In [8]:
df=df1.dropna(subset=["price"], axis=0)
df.head(10)

symboling normalized-losses         make  ... city-mpg highway-mpg  price
0           3               NaN  alfa-romero  ...       21          27  13495
1           3               NaN  alfa-romero  ...       21          27  16500
2           1               NaN  alfa-romero  ...       19          26  16500
3           2               164         audi  ...       24          30  13950
4           2               164         audi  ...       18          22  17450
5           2               NaN         audi  ...       19          25  15250
6           1               158         audi  ...       19          25  17710
7           1               NaN         audi  ...       19          25  18920
8           1               158         audi  ...       17          20  23875
10          2               192          bmw  ...       23          29  16430

[10 rows x 26 columns]

**Find the name of the columns of the dataframe.**

In [9]:
print(df.columns)

Index(['symboling', 'normalized-losses', 'make', 'fuel-type', 'aspiration',
       'num-of-doors', 'body-style', 'drive-wheels', 'engine-location',
       'wheel-base', 'length', 'width', 'height', 'curb-weight', 'engine-type',
       'num-of-cylinders', 'engine-size', 'fuel-system', 'bore', 'stroke',
       'compression-ratio', 'horsepower', 'peak-rpm', 'city-mpg',
       'highway-mpg', 'price'],
      dtype='object')


## Save Dataset
Correspondingly, Pandas enables us to save the dataset to csv. By using the dataframe.to_csv() method, you can add the file path and name along with quotation marks in the brackets.

For example, if you would save the dataframe df as automobile.csv to your local machine, you may use the syntax below, where index = False means the row names will not be written.

In [10]:
df.to_csv("automobile.csv", index=False)

## Basic Insight of Dataset
After reading data into Pandas dataframe, it is time for us to explore the dataset.
There are several ways to obtain essential insights of the data to help us better understand our dataset.

## Data Types
Data has a variety of types.
The main types stored in Pandas dataframes are object, float, int, bool and datetime64. In order to better learn about each attribute, it is always good for us to know the data type of each column. In Pandas:

In [11]:
df.dtypes

symboling              int64
normalized-losses     object
make                  object
fuel-type             object
aspiration            object
num-of-doors          object
body-style            object
drive-wheels          object
engine-location       object
wheel-base           float64
length               float64
width                float64
height               float64
curb-weight            int64
engine-type           object
num-of-cylinders      object
engine-size            int64
fuel-system           object
bore                  object
stroke                object
compression-ratio    float64
horsepower            object
peak-rpm              object
city-mpg               int64
highway-mpg            int64
price                 object
dtype: object

## Describe
If we would like to get a statistical summary of each column e.g. count, column mean value, column standard deviation, etc., we use the describe method:

In [12]:
df.describe()

symboling  wheel-base  ...    city-mpg  highway-mpg
count  201.000000  201.000000  ...  201.000000   201.000000
mean     0.840796   98.797015  ...   25.179104    30.686567
std      1.254802    6.066366  ...    6.423220     6.815150
min     -2.000000   86.600000  ...   13.000000    16.000000
25%      0.000000   94.500000  ...   19.000000    25.000000
50%      1.000000   97.000000  ...   24.000000    30.000000
75%      2.000000  102.400000  ...   30.000000    34.000000
max      3.000000  120.900000  ...   49.000000    54.000000

[8 rows x 10 columns]

This shows the statistical summary of all numeric-typed (int, float) columns.
For example, the attribute "symboling" has 205 counts, the mean value of this column is 0.83, the standard deviation is 1.25, the minimum value is -2, 25th percentile is 0, 50th percentile is 1, 75th percentile is 2, and the maximum value is 3.

However, what if we would also like to check all the columns including those that are of type object?


You can add an argument **include = "all"** inside the bracket. Let's try it again.

In [13]:
# describe all the columns in "df" 
df.describe(include = "all")

symboling normalized-losses    make  ...    city-mpg highway-mpg price
count   201.000000               164     201  ...  201.000000  201.000000   201
unique         NaN                51      22  ...         NaN         NaN   186
top            NaN               161  toyota  ...         NaN         NaN  7609
freq           NaN                11      32  ...         NaN         NaN     2
mean      0.840796               NaN     NaN  ...   25.179104   30.686567   NaN
std       1.254802               NaN     NaN  ...    6.423220    6.815150   NaN
min      -2.000000               NaN     NaN  ...   13.000000   16.000000   NaN
25%       0.000000               NaN     NaN  ...   19.000000   25.000000   NaN
50%       1.000000               NaN     NaN  ...   24.000000   30.000000   NaN
75%       2.000000               NaN     NaN  ...   30.000000   34.000000   NaN
max       3.000000               NaN     NaN  ...   49.000000   54.000000   NaN

[11 rows x 26 columns]

Apply the method to ".describe()" to the columns 'length' and 'compression-ratio'.

In [14]:
# Write your code below and press Shift+Enter to execute 
df[['length' , 'compression-ratio']].describe()

length  compression-ratio
count  201.000000         201.000000
mean   174.200995          10.164279
std     12.322175           4.004965
min    141.100000           7.000000
25%    166.800000           8.600000
50%    173.200000           9.000000
75%    183.500000           9.400000
max    208.100000          23.000000

## Info
Another method you can use to check your dataset is:

It provides a concise summary of your DataFrame.

This method prints information about a DataFrame including the index dtype and columns, non-null values and memory usage.

In [15]:
# look at the info of "df"
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 201 entries, 0 to 204
Data columns (total 26 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   symboling          201 non-null    int64  
 1   normalized-losses  164 non-null    object 
 2   make               201 non-null    object 
 3   fuel-type          201 non-null    object 
 4   aspiration         201 non-null    object 
 5   num-of-doors       199 non-null    object 
 6   body-style         201 non-null    object 
 7   drive-wheels       201 non-null    object 
 8   engine-location    201 non-null    object 
 9   wheel-base         201 non-null    float64
 10  length             201 non-null    float64
 11  width              201 non-null    float64
 12  height             201 non-null    float64
 13  curb-weight        201 non-null    int64  
 14  engine-type        201 non-null    object 
 15  num-of-cylinders   201 non-null    object 
 16  engine-size        201 non